File: template-ts/src/sim/replay.ts
/**
 * Deterministic replay: the template's most load-bearing test primitive.
 *
 * A replay is `(seed, per-tick intents)`. Running it produces a per-tick hash
 * chain. Two runs of the same replay must produce identical chains; if they do
 * not, something in the simulation is order-dependent, clock-dependent, or
 * reading unseeded entropy.
 *
 * This single mechanism catches most determinism regressions, which is why
 * every gameplay change should come with a replay test.
 */

import {
  type Intents,
  NO_INTENTS,
  type Score,
  type Seed,
  type World,
  spawnWorld,
  stateHash,
  step,
} from './index.ts';

/** A recorded run: everything needed to reproduce a simulation exactly. */
export interface Replay {
  readonly seed: Seed;
  /** Intent for each tick, in order. Length determines the run length. */
  readonly inputs: readonly Intents[];
}

export function replay(seed: Seed, inputs: readonly Intents[]): Replay {
  return { seed, inputs };
}

/** A replay with no player input — useful for testing the ball alone. */
export function idleReplay(seed: Seed, ticks: number): Replay {
  return { seed, inputs: new Array<Intents>(ticks).fill(NO_INTENTS) };
}

/** Outcome of running a replay. */
export interface ReplayOutcome {
  /** World hash after each tick. `hashes[i]` is the state after tick `i + 1`. */
  readonly hashes: readonly bigint[];
  readonly finalTick: number;
  readonly finalScore: Score;
}

/** Hash of the whole run — cheap to compare and to store as a golden value. */
export function digest(outcome: ReplayOutcome): bigint {
  const PRIME = 0x00000100000001b3n;
  const MASK = (1n << 64n) - 1n;
  let acc = 0xcbf29ce484222325n;
  for (const hash of outcome.hashes) {
    acc = ((acc ^ hash) * PRIME) & MASK;
  }
  return acc;
}

/**
 * Build a headless world.
 *
 * There is nothing to warm up and no event loop to drive: the invariant is
 * exact by construction — **after `headlessWorld`, one `step()` is one tick.**
 * Tests advance time by calling `step`, never by elapsed seconds.
 */
export function headlessWorld(seed: Seed): World {
  return spawnWorld(seed);
}

/** Run a replay to completion, hashing the world after every tick. */
export function run(replay: Replay): ReplayOutcome {
  const world = headlessWorld(replay.seed);
  const hashes: bigint[] = [];

  for (const intents of replay.inputs) {
    step(world, intents);
    hashes.push(stateHash(world));
  }

  return {
    hashes,
    finalTick: world.tick,
    finalScore: { ...world.score },
  };
}

/**
 * Run the same replay twice and return the first tick at which the two runs
 * diverge, or `null` if they are identical.
 *
 * This is the assertion behind `just test-determinism`. It is deliberately
 * exact: any divergence at all is a bug, not a tolerance to be widened.
 */
export function findDivergence(replay: Replay): number | null {
  const a = run(replay);
  const b = run(replay);
  const shared = Math.min(a.hashes.length, b.hashes.length);
  for (let tick = 0; tick < shared; tick += 1) {
    if (a.hashes[tick] !== b.hashes[tick]) {
      return tick;
    }
  }
  return a.hashes.length === b.hashes.length ? null : shared;
}

File: template-unity/Assets/Sim/Replay.cs
// Deterministic replay: the template's most load-bearing test primitive.
//
// A replay is (seed, per-tick intents). Running it produces a per-tick hash
// chain. Two runs of the same replay must produce identical chains; if they do
// not, something in the simulation is order-dependent, clock-dependent, or
// reading unseeded entropy.
//
// This single mechanism catches most determinism regressions, which is why
// every gameplay change should come with a replay test.

using System;
using System.Collections.Generic;

namespace Pong.Sim
{
    /// A recorded run: everything needed to reproduce a simulation exactly.
    public sealed class Replay
    {
        public readonly ulong Seed;

        /// Intent for each tick, in order. Length determines the run length.
        public readonly Intents[] Inputs;

        public Replay(ulong seed, Intents[] inputs)
        {
            Seed = seed;
            Inputs = inputs ?? Array.Empty<Intents>();
        }

        /// A replay with no player input — useful for testing the ball alone.
        public static Replay Idle(ulong seed, int ticks) => new Replay(seed, new Intents[ticks]);

        /// A replay that holds the same intent for every tick.
        public static Replay Held(ulong seed, Intents intent, int ticks)
        {
            var inputs = new Intents[ticks];
            for (int i = 0; i < ticks; i++) inputs[i] = intent;
            return new Replay(seed, inputs);
        }

        public int Length => Inputs.Length;
        public bool IsEmpty => Inputs.Length == 0;

        public Replay Prefix(int ticks)
        {
            var slice = new Intents[ticks];
            Array.Copy(Inputs, slice, ticks);
            return new Replay(Seed, slice);
        }

        /// Run a replay to completion, hashing the world after every tick.
        public static ReplayOutcome Run(Replay replay)
        {
            var state = new SimState(replay.Seed);
            var hashes = new ulong[replay.Length];
            for (int i = 0; i < replay.Length; i++)
            {
                state.Step(replay.Inputs[i]);
                hashes[i] = state.StateHash();
            }
            return new ReplayOutcome(hashes, state.Tick, state.Score);
        }

        /// Run the same replay twice and return the first tick at which the two
        /// runs diverge, or -1 if they are identical.
        ///
        /// This is deliberately exact: any divergence at all is a bug, not a
        /// tolerance to be widened.
        public static int FindDivergence(Replay replay)
        {
            var a = Run(replay);
            var b = Run(replay);
            for (int i = 0; i < a.Hashes.Count && i < b.Hashes.Count; i++)
            {
                if (a.Hashes[i] != b.Hashes[i]) return i;
            }
            return -1;
        }
    }

    /// Outcome of running a replay.
    public sealed class ReplayOutcome
    {
        /// World hash after each tick. `Hashes[i]` is the state after tick i+1.
        public readonly IReadOnlyList<ulong> Hashes;
        public readonly ulong FinalTick;
        public readonly Score FinalScore;

        public ReplayOutcome(ulong[] hashes, ulong finalTick, Score finalScore)
        {
            Hashes = hashes;
            FinalTick = finalTick;
            FinalScore = finalScore;
        }

        /// Hash of the whole run — cheap to compare and to store as a golden.
        public ulong Digest()
        {
            const ulong PRIME = 0x00000100000001b3UL;
            ulong acc = 0xcbf29ce484222325UL;
            foreach (var h in Hashes) acc = unchecked((acc ^ h) * PRIME);
            return acc;
        }
    }
}
